File: GO_ogle/wordSquares.js
class TrieNode {
  constructor() {
    this.children = {};
    this.isWord = false;
  }
}

class WordSquares {
  constructor() {
    this.root = new TrieNode();
  }

  wordSquares(words) {
    let root = this.buildTrie(words);
    let squares = [];

    for (let word of words) {
      let square = [];
      square.push(word);
      this.wordSquaresLong(root, word.length, square, squares);
    }
    return squares;
  }

  buildTrie(words) {
    let root = this.root;

    for (let word of words) {
      let current = root;
        
      for (let char of word) {
        if (!current.children.hasOwnProperty(char)) {
          current.children[char] = new TrieNode();
        }
        current = current.children[char];
      }
      current.isWord = true;
    }
    return root;
  }

  search(root, prefix) {
    let current = root;
    for (let char of prefix) {
      if (!current.children.hasOwnProperty(char)) {
        return null;
      }
      current = current.children[char];
    }
    return current;
  }

  wordSquaresLong(root, len, square, squares) {
    if (square.length === len) {
      squares.push(square.slice());
      return;
    }
    let prefix = this.getPrefix(square, square.length);
    let node = this.search(root, prefix);
    if (!node) {
      return;
    }

    let children = [];
    this.getChildren(node, prefix, children);

    for (let child of children) {
      square.push(child);
      this.wordSquaresLong(root, len, square, squares);
      square.splice(square.length - 1, 1);
    }
  }

  getPrefix(square, index) {
    let prefix = '';
    for (let i = 0; i < index; i++) {
      prefix += square[i][index];
    }
    return prefix;
  }

  getChildren(node, str, children) {
    if (node.isWord) {
      children.push(str);
      return;
    }

    for (let child of Object.keys(node.children)) {
      this.getChildren(node.children[child], str + child, children);
    }
  }
}


/**
 * @param {string[]} words
 * @return {string[][]}
 */
var wordSquares = function(words) {
    let wordSquaresClass = new WordSquares();
    return wordSquaresClass.wordSquares(words);
};

wordSquares(['ball', 'lady', 'lead', 'area']);

File: GO_ogle/wordSquares.scala
import scala.collection.mutable.ListBuffer
import scala.collection.mutable.HashMap
object Solution {
  def wordSquares(words: Array[String]): List[List[String]] = {
    val result = ListBuffer[ListBuffer[String]]()
    val map = createPrefixMap(words)
    for (i <- 0 until words.length) {
      val cList = ListBuffer[String](words(i))
      backTrack(1, cList, result, words(i).length, map)
    }

    result.map(_.toList).toList
  }

  private def backTrack(
      step: Int,
      list: ListBuffer[String],
      result: ListBuffer[ListBuffer[String]],
      len: Int,
      map: HashMap[String, ListBuffer[String]]
  ): Unit = {
    if (list.size == len) {
      result += list.clone
      return ()
    }
    val sb = new StringBuilder()
    for (word <- list) sb.append(word(step))
    val prefix = sb.toString()
    val wordList = map.getOrElse(prefix, ListBuffer[String]())

    for (word <- wordList) {
      list += word
      backTrack(step + 1, list, result, len, map)
      list.remove(list.size - 1)
    }
  }

  private def createPrefixMap(
      words: Array[String]
  ): HashMap[String, ListBuffer[String]] = {
    words.foldLeft(HashMap[String, ListBuffer[String]]()) {
      case (map, word) => {
        for (i <- 0 until word.length) {
          val prefix = word.substring(0, i)
          map(prefix) = map.getOrElse(prefix, ListBuffer[String]()) :+ word
        }
        map
      }
    }
  }
}

/*
G: words: Array[String]
O: allWordSquares: List[List[String]]
T: O(N⋅26^L) where L is the length of a single word
S: O(N⋅L) where L is the length of a single word

Notes:
  - Given a set of words (without duplicates)
  - There are at least 1 and at most 1000 words.
  - All words will have the exact same length.
  - Word length is at least 1 and at most 5.
  - Each word contains only lowercase English alphabet a-z.

Ex:
Input:
["area","lead","wall","lady","ball"]

Output:
[
  [ "wall",
    "area",
    "lead",
    "lady"
  ],
  [ "ball",
    "area",
    "lead",
    "lady"
  ]
]

Explanation:
The output consists of two word squares. The order of output does not matter (just the order of words in each word square matters).

 */
